src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchOperations.java

```diff
@@ -406,4 +407,26 @@ default String stringIdRepresentation(@Nullable Object id) {
 		return Objects.toString(id, null);
 	}
 	// endregion
+
+	/**
+	 * Execute a callback with the {@link RestHighLevelClient}
+	 * 
+	 * @param callback the callback to execute, must not be {@literal null}
+	 * @param <T> the type returned from the callback
+	 * @return the callback result
+	 * @throws UnsupportedOperationException for the TransportClient based implementation.
+	 * @since 4.0
+	 */
+	<T> T execute(RestHighLevelClientCallback<T> callback) throws RuntimeException;
+
+	/**
+	 * Callback interface to be used with {@link #execute(RestHighLevelClientCallback)} for operating directly on
+	 * {@link org.springframework.data.elasticsearch.client.reactive.ReactiveElasticsearchClient}.
+	 * 
+	 * @since 4.0
+	 */
+	@FunctionalInterface
+	interface RestHighLevelClientCallback<T> {
+		T doWithClient(RestHighLevelClient client) throws Throwable;
```

Comment:
Declaring a checked exception such as `IOException` is fine but we must not generalize towards `Throwable`.